src/Adapter/MSTest.TestAdapter/Discovery/AssemblyEnumerator.cs

```diff
@@ -384,100 +352,142 @@ static UnitTestElement GetFixtureTest(string classFullName, string assemblyLocat
         }
     }
 
-    private static bool TryProcessITestDataSourceTests(UnitTestElement test, TestMethodInfo testMethodInfo, List<UnitTestElement> tests)
+    private static bool TryUnfoldITestDataSources(UnitTestElement test, TestMethodInfo testMethodInfo, TestDataSourceUnfoldingStrategy dataSourcesUnfoldingStrategy, List<UnitTestElement> tests)
     {
+        // It should always be `true`, but if any part of the chain is obsolete; it might not contain those.
+        // Since we depend on those properties, if they don't exist, we bail out early.
+        if (!test.TestMethod.HasManagedMethodAndTypeProperties)
+        {
+            return false;
+        }
+
         // We don't have a special method to filter attributes that are not derived from Attribute, so we take all
         // attributes and filter them. We don't have to care if there is one, because this method is only entered when
         // there is at least one (we determine this in TypeEnumerator.GetTestFromMethod.
         IEnumerable<ITestDataSource> testDataSources = ReflectHelper.Instance.GetDerivedAttributes<Attribute>(testMethodInfo.MethodInfo, inherit: false).OfType<ITestDataSource>();
 
+        // We need to use a temporary list to avoid adding tests to the main list if we fail to expand any data source.
+        List<UnitTestElement> tempListOfTests = new();
+
         try
         {
-            return ProcessITestDataSourceTests(test, new(testMethodInfo.MethodInfo, test.DisplayName), testDataSources, tests);
+            bool isDataDriven = false;
+            foreach (ITestDataSource dataSource in testDataSources)
+            {
+                isDataDriven = true;
+                if (!TryUnfoldITestDataSource(dataSource, dataSourcesUnfoldingStrategy, test, new(testMethodInfo.MethodInfo, test.DisplayName), tempListOfTests))
+                {
+                    // TODO: Improve multi-source design!
```

Comment:
Yes as it would force me to have some kind of identifier of the data source so that I know the ones to exclude and there is no good way to do that for now.